game/game/movealgo.cpp

```diff
@@ -708,6 +708,38 @@ void MoveAlgo::startDive() {
     }
   }
 
+void MoveAlgo::invalidatePhysics(const Tempest::Vec3& dp) {
+  auto  pos           = npc.position();
+  float fallThreshold = stepHeight();
+  float worldHeight   = DynamicWorld::worldHeight;
+  auto  pathRay       = npc.world().physic()->ray(pos-dp,pos);
+  auto  skyRay        = npc.world().physic()->ray(Tempest::Vec3(pos.x,worldHeight,pos.z),pos);
+  bool  validW        = false;
+  waterRay(pos,&validW);
+
+  if(pathRay.hasCol) {
+    if (skyRay.hasCol)
+      npc.setPosition(skyRay.v+Tempest::Vec3(0,5*fallThreshold,0));
+    setAsSwim(false);
+    setAsDive(false);
+    setInAir (true);
+    return;
+    }
+
+  if (!validW) {
+    auto groundRay = npc.world().physic()->ray(pos,Tempest::Vec3(pos.x,-worldHeight,pos.z));
```

Comment:
Why not `landRay`? Also why sometimes you reaching to `physics()` directly and sometimes, thru cache, like with `waterRay`?